[issue/17] - Add sections, [issue/20] - Add Makefile, [issue/20] - Add styles (#30)
teamulster2/report: report/04_implementation/01_flutter_application.md

Because of this, the development does not take place platform-specifically,
but in one programming language and environment.

The programming language used with *Flutter* is *Dart*. 
*Dart* is an object oriented programming language that is very similar to C#
or Java. 

It's advantage over Java, C# etc. is its modern implementation where it fixes 

> common issues of old languages like Java. Dart for example implements null 

Author: I'd rather omit this suggestion since elder or elderly is mostly used when referring to old people or antique objects. Imo Java is not yet antique but it isn't quite far away from becoming so 😁

> common issues of old languages like Java. Dart for example implements null 

Reviewer: If the name doesn't fit you, that's okay. I know elder as a more careful way of expressing something "old", but I could be wrong about that. The repetition of "Java" bothered me more than the old ...

> common issues of old languages like Java. Dart for example implements null 

Author: Alright, then - I see.
In my next commit I rewrote the whole sentence, so that it no longer states `Java` twice. 👍🏿

    // The unnamed method the init function receives will set the variable 
    // _doneInitializing to true which allows the typical Flutter build method 
    // to ensure that every key has been initialized and returns the actual 
    // app.
    Provider.of<ServiceProvider>(context, listen: false)
        .init(() => setState(() => _doneInitializing = true), path);
  }

  @override
  Widget build(BuildContext context) {
    if (_doneInitializing) {
      return MaterialApp(...);
    }
  }

  ...

}

...

class ServiceProvider extends ChangeNotifier {
  final ConfigManager _configManager = ConfigManager();
  final DatabaseManager _databaseManager = DatabaseManager();
  final Notifications _notifications = Notifications();

  // ...
  // Getter and Setter
  // ...

  Future<void> init(Function onDoneInitializing, <more arguments> ...) async {
    // ...
    // initialization tasks
    // ...

    onDoneInitializing(); // <- invoke this method after all init tasks have 
                          // completed
  }
}
```

### Configuration Manager

As mentioned in [Service Provider](service-provider) the configuration manager 

> is part of it. It also implements the Singleton design pattern principle to 

Reviewer: part of what?
```suggestion
is part of the service provider. It also implements the Singleton design pattern principle to 
```

Author: Since this sentence already contains a reference to the service provider, I think it is clear that `it` is referring to the service provider. Therefore, I'd prefer omitting this suggestion as well 😅

Reviewer: The only problem I had while reading was that I had to read twice what was meant. But yeah, I got it after a while.

> ensure only one instance of itself and only one instance of client config 

Reviewer: ```suggestion
ensure, that only one instance of itself and only one instance of client config 
```

exists. The config manager provides various functions to write and read data 
to / from a local json file, although its main purpose is fetching a 
configuration from the study server. If no server can be reached and / or no 
local json file exists, it holds a default configuration which can be loaded 
any time.

### Database Manager

The database package is a relatively big package, as it holds multiple models 
and their generated adapters. As every key element of the app it is also part 
of the service provider class.  

> As most databases it contains basic CRUD (Create, Read, Update, Delete) 

Reviewer: ```suggestion
As most databases, it contains basic CRUD (Create, Read, Update, Delete) 
```

operations and also provides the ability to export all database contents as 

> json, so it can be send to the server. Mentionable when using *Hive* as 

Reviewer: ```suggestion
json, so they can be send to the server. Mentionable when using *Hive* as 
```

Author: Imo `it` is a better fit, because it is referring to one json object / file containing all database entries 🤔

database is the fact that *Hive* automatically updates an entry if another 
entry with matching keys is written to the same box (can be compared to tables 

> in relational database).

Reviewer: ```suggestion
in relational database or collections in well-known NoSQL databases).
```

### Exceptions

> Exception handling in the soTired is quite straight forward. While developing 

Reviewer: ```suggestion
Exception handling in the soTired app is quite straight forward. While developing 
```

> multiple custom exception have been implemented for specific use cases. When 

Reviewer: ```suggestion
multiple custom exceptions have been implemented for specific use cases. When 
```

Because of this, the development does not take place platform-specifically,
but in one programming language and environment.

The programming language used with *Flutter* is *Dart*. 
*Dart* is an object oriented programming language that is very similar to C#
or Java. 

Its advantage over Java, C# etc. is a modern implementation where common 
issues have been fixed, that older languages weren't aware of while being 

> developed. Dart for example implements null sound safety null checker where 

Reviewer 2: ```suggestion
developed. Dart for example implements null sound safety checker where 
```

logic.

### Service Provider

The key part of the app is provided by a service provider. This service 
provider ensures that all key elements like the config manager, the database 
manager and notifications have been initialized successfully before enabling 
the ui and allowing the user to interact with the app.  
During runtime of the app the service provider can be accessed from everywhere 
within the app's source code with help from the 
[provider](https://pub.dev/packages/provider) package. This has the advantage 
of utilizing the Singleton design pattern principle for all key parts of the 
app.  
While the provider package ensures the existence of only one service provider 
instance, a critical issue arises when initializing all key parts. This is 
mostly because of the asynchronous execution of code parts which have to be 
synchronized. Solving this issue needed some sort of *Mutex* to be 
implemented.  
Unfortunately, *Dart* is kind of restrictive when it comes to using 
rudimentary programming principles like *Mutexes*. Therefore, the `init` 
method of the service provider takes a so called `Closure` which can be 

> described as nameless function. This function is invoked after all 
> initialization tasks have been completed.  

Reviewer 2: ```suggestion
described as function with access to variable/value beyond her scope.
This function is invoked after all initialization tasks have been completed.  
```

The code example below explains this process in a more detailed manner.

> ```dart

Reviewer 2: I think since the code is hear meant to be read by not programmers it would be kind to add more comments explaining more what is going on in non programmer understandable text :)

Author: Imo more comments contradict the fact of adding code examples to this report. Since the basic functionality is already explained as plain text this code example is only an enhancement for programmers who are interested in further understanding how this can be implemented in Dart.

In short, non programmers will most likely skip the code example ^^

Reviewer 2: sure if there are no comments that help them to understand it ;)
Also IMHO your last comment contradicts the line above the code example:
```
The code example below explains this process in a more detailed manner.
```
Since you only show the code and don't explain it in more detail.

be carried out once and not several times. In addition, the individual areas
should be easy to find by other programmers.

### Testing

> Testing within soTired was considered as important as every other part of the 

Reviewer: ```suggestion
Testing within soTired was considered as very important like every other part of the 
```

Reviewer: You can also use 'considered as as important as every ...', but that sounds a little bit weird. For me in the orginal statement one 'as' is missing..

Author: `as ... as` is commonly used to compare things. Therefore, I think I have to insist on leaving this untouched 😅

be carried out once and not several times. In addition, the individual areas
should be easy to find by other programmers.

### Testing

Testing within soTired was considered as important as every other part of the 
app. Every function of the config manager and the database manager is 

> rigorously unit tested which ensures a certain amount of code quality.  

Reviewer 2: ```suggestion
rigorously unit tested which ensures a certain amount of confidence when changing code.  
```

Author: I applied both because I think code quality is also ensured with the help of tests 👍🏿
